File: src/c/scheduler.c
/*
 * CPU Scheduling Visualizer - Core Scheduler Implementation
 * Implements the scheduling engine with support for all algorithms
 */

#include "scheduler.h"
#include <stdio.h>
#include <string.h>

/* ===================== Queue Operations ===================== */

void queue_init(ReadyQueue *q) {
  q->front = 0;
  q->rear = -1;
  q->size = 0;
  memset(q->pids, -1, sizeof(q->pids));
}

void queue_push(ReadyQueue *q, int pid) {
  if (q->size >= MAX_PROCESSES)
    return;
  q->rear = (q->rear + 1) % MAX_PROCESSES;
  q->pids[q->rear] = pid;
  q->size++;
}

int queue_pop(ReadyQueue *q) {
  if (q->size == 0)
    return -1;
  int pid = q->pids[q->front];
  q->front = (q->front + 1) % MAX_PROCESSES;
  q->size--;
  return pid;
}

int queue_peek(ReadyQueue *q) {
  if (q->size == 0)
    return -1;
  return q->pids[q->front];
}

int queue_is_empty(ReadyQueue *q) { return q->size == 0; }

int queue_size(ReadyQueue *q) { return q->size; }

void queue_remove(ReadyQueue *q, int pid) {
  /* Remove a specific PID from queue (for preemption) */
  int new_pids[MAX_PROCESSES];
  int new_size = 0;

  for (int i = 0; i < q->size; i++) {
    int idx = (q->front + i) % MAX_PROCESSES;
    if (q->pids[idx] != pid) {
      new_pids[new_size++] = q->pids[idx];
    }
  }

  /* Rebuild queue */
  queue_init(q);
  for (int i = 0; i < new_size; i++) {
    queue_push(q, new_pids[i]);
  }
}

/* ===================== Scheduler Core ===================== */

void scheduler_init(Scheduler *s) {
  s->process_count = 0;
  s->current_time = 0;
  s->running_pid = -1;
  s->time_quantum = 2;
  s->quantum_remaining = 0;
  s->algorithm = ALGO_FCFS;
  s->num_cores = 1;
  s->is_completed = 0;
  s->gantt_count = 0;
  s->starvation_threshold = 10;
  s->aging_enabled = 0;
  s->aging_interval = 5;

  queue_init(&s->ready_queue);

  for (int i = 0; i < MAX_CORES; i++) {
    s->running_pids[i] = -1;
  }

  memset(&s->metrics, 0, sizeof(Metrics));
}

void scheduler_reset(Scheduler *s) {
  /* Reset simulation state but keep processes */
  s->current_time = 0;
  s->running_pid = -1;
  s->quantum_remaining = 0;
  s->is_completed = 0;
  s->gantt_count = 0;

  queue_init(&s->ready_queue);

  for (int i = 0; i < MAX_CORES; i++) {
    s->running_pids[i] = -1;
  }

  /* Reset process states */
  for (int i = 0; i < s->process_count; i++) {
    s->processes[i].remaining_time = s->processes[i].burst_time;
    s->processes[i].start_time = -1;
    s->processes[i].finish_time = -1;
    s->processes[i].wait_time = 0;
    s->processes[i].response_time = -1;
    s->processes[i].turnaround_time = 0;
    s->processes[i].state = STATE_NEW;
  }

  memset(&s->metrics, 0, sizeof(Metrics));
}

void scheduler_set_algorithm(Scheduler *s, Algorithm algo) {
  s->algorithm = algo;
}

void scheduler_set_time_quantum(Scheduler *s, int quantum) {
  s->time_quantum = quantum > 0 ? quantum : 1;
}

void scheduler_set_cores(Scheduler *s, int num_cores) {
  s->num_cores = (num_cores > 0 && num_cores <= MAX_CORES) ? num_cores : 1;
}

/* ===================== Process Management ===================== */

int scheduler_add_process(Scheduler *s, int arrival_time, int burst_time,
                          int priority) {
  if (s->process_count >= MAX_PROCESSES)
    return -1;

  int pid = s->process_count;
  init_process(&s->processes[pid], pid, arrival_time, burst_time, priority);
  s->process_count++;

  return pid;
}

Process *scheduler_get_process(Scheduler *s, int pid) {
  if (pid < 0 || pid >= s->process_count)
    return NULL;
  return &s->processes[pid];
}

int scheduler_get_process_index(Scheduler *s, int pid) {
  for (int i = 0; i < s->process_count; i++) {
    if (s->processes[i].pid == pid)
      return i;
  }
  return -1;
}

/* ===================== Algorithm Selection ===================== */

int select_process_fcfs(Scheduler *s) {
  /* FCFS: Select process that arrived first (front of queue) */
  if (queue_is_empty(&s->ready_queue))
    return -1;
  return queue_peek(&s->ready_queue);
}

int select_process_sjf(Scheduler *s) {
  /* SJF: Select process with smallest burst time */
  if (queue_is_empty(&s->ready_queue))
    return -1;

  int selected_pid = -1;
  int min_burst = 999999;

  for (int i = 0; i < s->ready_queue.size; i++) {
    int idx = (s->ready_queue.front + i) % MAX_PROCESSES;
    int pid = s->ready_queue.pids[idx];
    Process *p = scheduler_get_process(s, pid);

    if (p && p->burst_time < min_burst) {
      min_burst = p->burst_time;
      selected_pid = pid;
    }
  }

  return selected_pid;
}

int select_process_srtf(Scheduler *s) {
  /* SRTF: Select process with smallest remaining time (preemptive) */
  if (queue_is_empty(&s->ready_queue))
    return -1;

  int selected_pid = -1;
  int min_remaining = 999999;

  for (int i = 0; i < s->ready_queue.size; i++) {
    int idx = (s->ready_queue.front + i) % MAX_PROCESSES;
    int pid = s->ready_queue.pids[idx];
    Process *p = scheduler_get_process(s, pid);

    if (p && p->remaining_time < min_remaining) {
      min_remaining = p->remaining_time;
      selected_pid = pid;
    }
  }

  return selected_pid;
}

int select_process_priority(Scheduler *s) {
  /* Priority: Select process with highest priority (lowest number) */
  if (queue_is_empty(&s->ready_queue))
    return -1;

  int selected_pid = -1;
  int highest_priority = 999999;

  for (int i = 0; i < s->ready_queue.size; i++) {
    int idx = (s->ready_queue.front + i) % MAX_PROCESSES;
    int pid = s->ready_queue.pids[idx];
    Process *p = scheduler_get_process(s, pid);

    if (p && p->priority < highest_priority) {
      highest_priority = p->priority;
      selected_pid = pid;
    }
  }

  return selected_pid;
}

int select_process_round_robin(Scheduler *s) {
  /* Round Robin: Take from front of queue */
  if (queue_is_empty(&s->ready_queue))
    return -1;
  return queue_peek(&s->ready_queue);
}

int select_process_ljf(Scheduler *s) {
  /* LJF: Select process with largest burst time */
  if (queue_is_empty(&s->ready_queue))
    return -1;

  int selected_pid = -1;
  int max_burst = -1;

  for (int i = 0; i < s->ready_queue.size; i++) {
    int idx = (s->ready_queue.front + i) % MAX_PROCESSES;
    int pid = s->ready_queue.pids[idx];
    Process *p = scheduler_get_process(s, pid);

    if (p && p->burst_time > max_burst) {
      max_burst = p->burst_time;
      selected_pid = pid;
    }
  }

  return selected_pid;
}

int select_process_lrtf(Scheduler *s) {
  /* LRTF: Select process with largest remaining time (preemptive) */
  if (queue_is_empty(&s->ready_queue))
    return -1;

  int selected_pid = -1;
  int max_remaining = -1;

  for (int i = 0; i < s->ready_queue.size; i++) {
    int idx = (s->ready_queue.front + i) % MAX_PROCESSES;
    int pid = s->ready_queue.pids[idx];
    Process *p = scheduler_get_process(s, pid);

    if (p && p->remaining_time > max_remaining) {
      max_remaining = p->remaining_time;
      selected_pid = pid;
    }
  }

  return selected_pid;
}

/* Select process based on current algorithm */
static int select_next_process(Scheduler *s) {
  switch (s->algorithm) {
  case ALGO_FCFS:
    return select_process_fcfs(s);
  case ALGO_SJF:
    return select_process_sjf(s);
  case ALGO_SRTF:
    return select_process_srtf(s);
  case ALGO_PRIORITY:
    return select_process_priority(s);
  case ALGO_ROUND_ROBIN:
    return select_process_round_robin(s);
  case ALGO_LJF:
    return select_process_ljf(s);
  case ALGO_LRTF:
    return select_process_lrtf(s);
  default:
    return select_process_fcfs(s);
  }
}

/* Check if algorithm is preemptive */
static int is_preemptive(Algorithm algo) {
  return (algo == ALGO_SRTF || algo == ALGO_LRTF || algo == ALGO_ROUND_ROBIN);
}

/* ===================== Gantt Chart ===================== */

static void add_gantt_entry(Scheduler *s, int pid, int start, int end,
                            int core_id) {
  if (s->gantt_count >= MAX_GANTT_ENTRIES)
    return;

  /* Try to merge with previous entry if same process */
  if (s->gantt_count > 0) {
    GanttEntry *last = &s->gantt[s->gantt_count - 1];
    if (last->pid == pid && last->end_time == start &&
        last->core_id == core_id) {
      last->end_time = end;
      return;
    }
  }

  GanttEntry *entry = &s->gantt[s->gantt_count++];
  entry->pid = pid;
  entry->start_time = start;
  entry->end_time = end;
  entry->core_id = core_id;
}

/* ===================== Simulation Step ===================== */

int scheduler_step(Scheduler *s) {
  if (s->is_completed)
    return 0;

  /* Step 1: Check for newly arrived processes (NEW -> READY) */
  for (int i = 0; i < s->process_count; i++) {
    Process *p = &s->processes[i];
    if (p->state == STATE_NEW && p->arrival_time <= s->current_time) {
      p->state = STATE_READY;
      queue_push(&s->ready_queue, p->pid);
    }
  }

  /* Step 2: Apply aging if enabled */
  if (s->aging_enabled && s->current_time > 0 &&
      s->current_time % s->aging_interval == 0) {
    scheduler_apply_aging(s);
  }

  /* Step 3: Handle preemption for preemptive algorithms */
  if (is_preemptive(s->algorithm) && s->running_pid != -1) {
    Process *running = scheduler_get_process(s, s->running_pid);

    if (s->algorithm == ALGO_ROUND_ROBIN) {
      /* Check quantum expiry */
      if (s->quantum_remaining <= 0) {
        /* Preempt: move running process back to ready queue */
        running->state = STATE_READY;
        queue_push(&s->ready_queue, s->running_pid);
        s->running_pid = -1;
      }
    } else {
      /* SRTF or LRTF: check if better process available */
      int next_pid = select_next_process(s);
      if (next_pid != -1 && next_pid != s->running_pid) {
        Process *next = scheduler_get_process(s, next_pid);
        int should_preempt = 0;

        if (s->algorithm == ALGO_SRTF) {
          should_preempt = next->remaining_time < running->remaining_time;
        } else if (s->algorithm == ALGO_LRTF) {
          should_preempt = next->remaining_time > running->remaining_time;
        }

        if (should_preempt) {
          running->state = STATE_READY;
          queue_push(&s->ready_queue, s->running_pid);
          s->running_pid = -1;
        }
      }
    }
  }

  /* Step 4: Select next process if CPU is idle */
  if (s->running_pid == -1) {
    int next_pid = select_next_process(s);

    if (next_pid != -1) {
      queue_remove(&s->ready_queue, next_pid);
      s->running_pid = next_pid;

      Process *p = scheduler_get_process(s, next_pid);
      p->state = STATE_RUNNING;

      /* Record first time process gets CPU */
      if (p->start_time == -1) {
        p->start_time = s->current_time;
        p->response_time = s->current_time - p->arrival_time;
      }

      /* Reset quantum for Round Robin */
      if (s->algorithm == ALGO_ROUND_ROBIN) {
        s->quantum_remaining = s->time_quantum;
      }
    }
  }

  /* Step 5: Execute one time unit */
  if (s->running_pid != -1) {
    Process *p = scheduler_get_process(s, s->running_pid);
    p->remaining_time--;

    if (s->algorithm == ALGO_ROUND_ROBIN) {
      s->quantum_remaining--;
    }

    /* Add to Gantt chart */
    add_gantt_entry(s, s->running_pid, s->current_time, s->current_time + 1, 0);

    /* Check if process completed */
    if (p->remaining_time <= 0) {
      p->state = STATE_TERMINATED;
      p->finish_time = s->current_time + 1;
      p->turnaround_time = p->finish_time - p->arrival_time;
      p->wait_time = p->turnaround_time - p->burst_time;
      s->running_pid = -1;
    }
  } else {
    /* CPU idle */
    add_gantt_entry(s, -1, s->current_time, s->current_time + 1, 0);
  }

  /* Step 6: Update waiting time for ready processes */
  for (int i = 0; i < s->ready_queue.size; i++) {
    int idx = (s->ready_queue.front + i) % MAX_PROCESSES;
    int pid = s->ready_queue.pids[idx];
    Process *p = scheduler_get_process(s, pid);
    if (p) {
      p->wait_time++;
    }
  }

  /* Step 7: Advance time */
  s->current_time++;

  /* Step 8: Check if simulation is complete */
  int all_terminated = 1;
  for (int i = 0; i < s->process_count; i++) {
    if (s->processes[i].state != STATE_TERMINATED) {
      all_terminated = 0;
      break;
    }
  }

  if (all_terminated && s->process_count > 0) {
    s->is_completed = 1;
    scheduler_calculate_metrics(s);
  }

  return !s->is_completed;
}

void scheduler_run_to_completion(Scheduler *s) {
  while (!s->is_completed && s->current_time < 10000) {
    scheduler_step(s);
  }
}

int scheduler_is_completed(Scheduler *s) { return s->is_completed; }

/* ===================== Metrics ===================== */

void scheduler_calculate_metrics(Scheduler *s) {
  if (s->process_count == 0)
    return;

  double total_wait = 0;
  double total_turnaround = 0;
  double total_response = 0;
  int total_burst = 0;
  int max_finish = 0;
  int min_arrival = 999999;

  for (int i = 0; i < s->process_count; i++) {
    Process *p = &s->processes[i];
    total_wait += p->wait_time;
    total_turnaround += p->turnaround_time;
    total_response += p->response_time;
    total_burst += p->burst_time;

    if (p->finish_time > max_finish)
      max_finish = p->finish_time;
    if (p->arrival_time < min_arrival)
      min_arrival = p->arrival_time;
  }

  s->metrics.avg_wait_time = total_wait / s->process_count;
  s->metrics.avg_turnaround_time = total_turnaround / s->process_count;
  s->metrics.avg_response_time = total_response / s->process_count;

  int total_time = max_finish - min_arrival;
  s->metrics.total_execution_time = total_time;
  s->metrics.total_idle_time = total_time - total_burst;

  if (total_time > 0) {
    s->metrics.cpu_utilization = (double)total_burst / total_time * 100.0;
    s->metrics.throughput = (double)s->process_count / total_time;
  }
}

/* ===================== Aging ===================== */

void scheduler_enable_aging(Scheduler *s, int interval) {
  s->aging_enabled = 1;
  s->aging_interval = interval > 0 ? interval : 5;
}

void scheduler_disable_aging(Scheduler *s) { s->aging_enabled = 0; }

void scheduler_apply_aging(Scheduler *s) {
  /* Boost priority of waiting processes */
  for (int i = 0; i < s->ready_queue.size; i++) {
    int idx = (s->ready_queue.front + i) % MAX_PROCESSES;
    int pid = s->ready_queue.pids[idx];
    Process *p = scheduler_get_process(s, pid);

    if (p && p->priority > 0) {
      p->priority--; /* Lower number = higher priority */
    }
  }
}

void scheduler_check_starvation(Scheduler *s) {
  for (int i = 0; i < s->ready_queue.size; i++) {
    int idx = (s->ready_queue.front + i) % MAX_PROCESSES;
    int pid = s->ready_queue.pids[idx];
    Process *p = scheduler_get_process(s, pid);

    if (p && p->wait_time >= s->starvation_threshold) {
      /* Process is starved - could trigger visual indicator */
    }
  }
}

File: backend/scheduler.py
# ================= CPU SCHEDULING ORCHESTRATOR =================
# All algorithm implementations live in backend/algorithms/
# This file wires them together for the tick-based API.

import sys
import os
sys.path.insert(0, os.path.dirname(__file__))

from algorithms import fcfs, sjf, srtf, priority_scheduling, round_robin, ljf, lrtf, mlfq

# =================== ALGORITHM MAP ===================

ALGORITHM_NAMES = {
    0: "FCFS",
    1: "SJF",
    2: "SRTF",
    3: "Priority",
    4: "Round Robin",
    5: "LJF",
    6: "LRTF",
    7: "MLFQ",
}

PREEMPTIVE = {2, 4, 6, 7}   # SRTF, Round Robin, LRTF, MLFQ

STATE_NAMES = {0: "NEW", 1: "READY", 2: "RUNNING", 3: "TERMINATED"}

# MLFQ queue quantums (Queue 0, Queue 1, Queue 2)
MLFQ_QUANTUMS = [4, 8, None]  # None = FCFS on lowest queue


# =================== TICK-BASED SCHEDULER (for frontend) ===================
# Selector functions decide which ready process runs next on each tick.

def _select_fcfs(ready_queue, processes, **kwargs):
    if not ready_queue: return -1
    return ready_queue[0]

def _select_sjf(ready_queue, processes, **kwargs):
    if not ready_queue: return -1
    return min(ready_queue, key=lambda pid: processes[pid]["burst_time"])

def _select_srtf(ready_queue, processes, **kwargs):
    if not ready_queue: return -1
    return min(ready_queue, key=lambda pid: processes[pid]["remaining_time"])

def _select_priority(ready_queue, processes, **kwargs):
    if not ready_queue: return -1
    return min(ready_queue, key=lambda pid: processes[pid]["priority"])

def _select_round_robin(ready_queue, processes, **kwargs):
    if not ready_queue: return -1
    return ready_queue[0]

def _select_ljf(ready_queue, processes, **kwargs):
    if not ready_queue: return -1
    return max(ready_queue, key=lambda pid: processes[pid]["burst_time"])

def _select_lrtf(ready_queue, processes, **kwargs):
    if not ready_queue: return -1
    return max(ready_queue, key=lambda pid: processes[pid]["remaining_time"])

def _select_mlfq(ready_queue, processes, mlfq_queues=None, **kwargs):
    """Pick the highest-priority non-empty MLFQ sub-queue's front process."""
    if mlfq_queues is None: return -1
    for q in mlfq_queues:
        if q:
            pid = q[0]
            if pid in ready_queue:
                return pid
    return -1

_ALGORITHM_FN = {
    0: _select_fcfs,
    1: _select_sjf,
    2: _select_srtf,
    3: _select_priority,
    4: _select_round_robin,
    5: _select_ljf,
    6: _select_lrtf,
    7: _select_mlfq,
}


def _make_process(pid, arrival, burst, priority):
    return {
        "pid": pid,
        "arrival_time": arrival,
        "burst_time": burst,
        "priority": priority,
        "remaining_time": burst,
        "start_time": -1,
        "finish_time": -1,
        "wait_time": 0,
        "response_time": -1,
        "turnaround_time": 0,
        "state": 0,  # 0=NEW, 1=READY, 2=RUNNING, 3=TERMINATED
        # MLFQ extras
        "mlfq_queue": 0,          # which MLFQ sub-queue the process is in
        "mlfq_quantum_used": 0,   # how many ticks used in current quantum slice
    }


class Scheduler:
    def __init__(self):
        self.processes = []
        self.ready_queue = []   # flat list of pids (for non-MLFQ algos)
        self.gantt = []
        self.current_time = 0
        self.running_pid = -1
        self.is_completed = False
        self.algorithm = 0
        self.time_quantum = 2
        self.quantum_remaining = 0
        self.metrics = {}

        # MLFQ-specific state: 3 sub-queues
        self.mlfq_queues = [[], [], []]

    def reset(self):
        self.current_time = 0
        self.running_pid = -1
        self.quantum_remaining = 0
        self.is_completed = False
        self.ready_queue = []
        self.gantt = []
        self.metrics = {}
        self.mlfq_queues = [[], [], []]
        for p in self.processes:
            p["remaining_time"] = p["burst_time"]
            p["start_time"] = -1
            p["finish_time"] = -1
            p["wait_time"] = 0
            p["response_time"] = -1
            p["turnaround_time"] = 0
            p["state"] = 0
            p["mlfq_queue"] = 0
            p["mlfq_quantum_used"] = 0

    def set_algorithm(self, algo):
        self.algorithm = algo

    def set_time_quantum(self, quantum):
        self.time_quantum = max(quantum, 1)

    def add_process(self, arrival, burst, priority=0):
        # Snap past arrival times to current time for dynamic additions
        if arrival < self.current_time:
            arrival = self.current_time

        pid = len(self.processes)
        self.processes.append(_make_process(pid, arrival, burst, priority))

        # If simulation was previously completed, allow resumption
        if self.is_completed:
            self.is_completed = False

        return pid

    def clear_processes(self):
        self.__init__()

    # ── One simulation tick ──

    def step(self):
        if self.is_completed:
            return False

        select_fn = _ALGORITHM_FN.get(self.algorithm, _select_fcfs)

        # ── 1. Admit newly arrived processes (NEW -> READY) ──
        for p in self.processes:
            if p["state"] == 0 and p["arrival_time"] <= self.current_time:
                p["state"] = 1
                if self.algorithm == 7:  # MLFQ: enter sub-queue 0, NOT flat ready_queue
                    p["mlfq_queue"] = 0
                    p["mlfq_quantum_used"] = 0
                    self.mlfq_queues[0].append(p["pid"])
                else:
                    self.ready_queue.append(p["pid"])

        # ── 2. Handle preemption ──
        if self.running_pid != -1:
            running = self.processes[self.running_pid]

            if self.algorithm == 7:  # MLFQ preemption
                q = running["mlfq_queue"]
                quantum = MLFQ_QUANTUMS[q]
                # Check if a higher-priority queue got a new arrival
                higher = any(self.mlfq_queues[qidx] for qidx in range(q))
                if higher:
                    # Preempt — put back at front of same MLFQ sub-queue
                    running["state"] = 1
                    self.mlfq_queues[q].insert(0, self.running_pid)
                    running["mlfq_quantum_used"] = 0  # reset slice
                    self.running_pid = -1
                elif quantum is not None and running["mlfq_quantum_used"] >= quantum:
                    # Quantum expired — demote
                    running["state"] = 1
                    next_q = min(q + 1, 2)
                    running["mlfq_queue"] = next_q
                    running["mlfq_quantum_used"] = 0
                    self.mlfq_queues[next_q].append(self.running_pid)
                    self.running_pid = -1

            elif self.algorithm == 4:  # Round Robin
                if self.quantum_remaining <= 0:
                    running["state"] = 1
                    self.ready_queue.append(self.running_pid)
                    self.running_pid = -1

            elif self.algorithm in PREEMPTIVE:  # SRTF / LRTF
                next_pid = select_fn(self.ready_queue, self.processes)
                if next_pid != -1 and next_pid != self.running_pid:
                    nxt = self.processes[next_pid]
                    should_preempt = False
                    if self.algorithm == 2:   # SRTF
                        should_preempt = nxt["remaining_time"] < running["remaining_time"]
                    elif self.algorithm == 6: # LRTF
                        should_preempt = nxt["remaining_time"] > running["remaining_time"]
                    if should_preempt:
                        running["state"] = 1
                        self.ready_queue.append(self.running_pid)
                        self.running_pid = -1

        # ── 3. Pick next process if CPU is idle ──
        if self.running_pid == -1:
            if self.algorithm == 7:
                # Pick front of highest-priority non-empty sub-queue
                next_pid = -1
                chosen_q = -1
                for qi, q in enumerate(self.mlfq_queues):
                    if q:
                        next_pid = q[0]
                        chosen_q = qi
                        break
                if next_pid != -1:
                    self.mlfq_queues[chosen_q].pop(0)
            else:
                next_pid = select_fn(self.ready_queue, self.processes)

            if next_pid != -1:
                if self.algorithm != 7:
                    self.ready_queue.remove(next_pid)
                self.running_pid = next_pid
                p = self.processes[next_pid]
                p["state"] = 2
                if p["start_time"] == -1:
                    p["start_time"] = self.current_time
                    p["response_time"] = self.current_time - p["arrival_time"]
                if self.algorithm == 4:   # RR
                    self.quantum_remaining = self.time_quantum
                if self.algorithm == 7:   # MLFQ: reset quantum counter
                    p["mlfq_quantum_used"] = 0

        # ── 4. Execute one time unit ──
        if self.running_pid != -1:
            p = self.processes[self.running_pid]
            p["remaining_time"] -= 1
            if self.algorithm == 4:
                self.quantum_remaining -= 1
            if self.algorithm == 7:
                p["mlfq_quantum_used"] += 1
            self._add_gantt(self.running_pid, self.current_time, self.current_time + 1)
            if p["remaining_time"] <= 0:
                p["state"] = 3
                p["finish_time"] = self.current_time + 1
                p["turnaround_time"] = p["finish_time"] - p["arrival_time"]
                p["wait_time"] = p["turnaround_time"] - p["burst_time"]
                self.running_pid = -1
        else:
            self._add_gantt(-1, self.current_time, self.current_time + 1)

        # ── 5. Increment waiting time for ready processes ──
        if self.algorithm == 7:
            # For MLFQ: processes in any sub-queue are waiting
            for q in self.mlfq_queues:
                for pid in q:
                    self.processes[pid]["wait_time"] += 1
        else:
            for pid in self.ready_queue:
                self.processes[pid]["wait_time"] += 1

        # ── 6. Advance clock ──
        self.current_time += 1

        # ── 7. Check completion ──
        if self.processes and all(p["state"] == 3 for p in self.processes):
            self.is_completed = True
            self._calculate_metrics()

        return not self.is_completed

    def run_to_completion(self):
        while not self.is_completed and self.current_time < 10000:
            self.step()

    # ── Gantt chart helper ──

    def _add_gantt(self, pid, start, end):
        if self.gantt:
            last = self.gantt[-1]
            if last["pid"] == pid and last["endTime"] == start:
                last["endTime"] = end
                return
        self.gantt.append({"pid": pid, "startTime": start, "endTime": end, "coreId": 0})

    # ── Metrics ──

    def _calculate_metrics(self):
        if not self.processes:
            return
        n = len(self.processes)
        total_wait = sum(p["wait_time"] for p in self.processes)
        total_tat = sum(p["turnaround_time"] for p in self.processes)
        total_resp = sum(p["response_time"] for p in self.processes)
        total_burst = sum(p["burst_time"] for p in self.processes)
        max_finish = max(p["finish_time"] for p in self.processes)
        min_arrival = min(p["arrival_time"] for p in self.processes)
        total_time = max_finish - min_arrival
        self.metrics = {
            "avgWaitTime": round(total_wait / n, 2),
            "avgTurnaroundTime": round(total_tat / n, 2),
            "avgResponseTime": round(total_resp / n, 2),
            "cpuUtilization": round((total_burst / total_time) * 100, 2) if total_time > 0 else 0,
            "throughput": round(n / total_time, 4) if total_time > 0 else 0,
            "totalIdleTime": total_time - total_burst,
            "totalExecutionTime": total_time,
        }

    # ── Get full state as a dict (for the API) ──

    def get_state(self):
        return {
            "currentTime": self.current_time,
            "runningPid": self.running_pid,
            "isCompleted": self.is_completed,
            "algorithm": self.algorithm,
            "timeQuantum": self.time_quantum,
            "processes": [
                {
                    "pid": p["pid"],
                    "arrivalTime": p["arrival_time"],
                    "burstTime": p["burst_time"],
                    "priority": p["priority"],
                    "remainingTime": p["remaining_time"],
                    "startTime": p["start_time"],
                    "finishTime": p["finish_time"],
                    "waitTime": p["wait_time"],
                    "responseTime": p["response_time"],
                    "turnaroundTime": p["turnaround_time"],
                    "state": p["state"],
                    "stateName": STATE_NAMES.get(p["state"], "UNKNOWN"),
                    "mlfqQueue": p.get("mlfq_queue", 0),
                }
                for p in self.processes
            ],
            "gantt": list(self.gantt),
            "readyQueue": list(self.ready_queue),
            "metrics": self.metrics or {
                "avgWaitTime": 0,
                "avgTurnaroundTime": 0,
                "avgResponseTime": 0,
                "cpuUtilization": 0,
                "throughput": 0,
                "totalIdleTime": 0,
                "totalExecutionTime": 0,
            },
        }


# ================= Example =================
if __name__ == "__main__":
    import copy

    process_list = [
        {'pid': 'P1', 'at': 0, 'bt': 7, 'priority': 2},
        {'pid': 'P2', 'at': 2, 'bt': 4, 'priority': 1},
        {'pid': 'P3', 'at': 4, 'bt': 1, 'priority': 3},
        {'pid': 'P4', 'at': 5, 'bt': 4, 'priority': 2},
    ]

    fcfs(copy.deepcopy(process_list))
    sjf(copy.deepcopy(process_list))
    srtf(copy.deepcopy(process_list))
    priority_scheduling(copy.deepcopy(process_list))
    round_robin(copy.deepcopy(process_list), quantum=2)
    ljf(copy.deepcopy(process_list))
    lrtf(copy.deepcopy(process_list))
    mlfq(copy.deepcopy(process_list))
